statsmodels/kernel_methods/bw_crossvalidation.py

```diff
@@ -0,0 +1,445 @@
+from __future__ import print_function, absolute_import, division
+import numpy as np
+from ..compat.python import range
+from scipy import optimize
+from ._grid_interpolation import GridInterpolator
+
+
+class LeaveOneOut(object):
+    """
+    Implement full LeaveOneOut method.
+
+    Parameters
+    ----------
+    data: tuple of ndarray
+        Data to process.
+    is_sel: tuple of bool
+        Of same length as data, indicate for each array if they are to be filtered or not.
+    npts: int
+        Number of points in the dataset
+    """
+    def __init__(self, data, is_sel, npts):
+        self.data = data
+        self.is_sel = is_sel
+        self.npts = npts
+
+    @property
+    def nb_tests(self):
+        return self.npts
+
+    def __iter__(self):
+        data = self.data
+        is_sel = self.is_sel
+        sel = np.ones((self.npts,), dtype=bool)
+        n = self.npts
+        for i in range(n):
+            sel[i] = False
+            yield (i, tuple(d[sel] if is_sel[i] else d for i, d in enumerate(data)))
+            sel[i] = True
+
+
+class LeaveOneOutSampling(object):
+    """
+    Implement a heureustic for the LeaveOneOut method in which only a sampling of the data are left out.
+
+    Parameters
+    ----------
+    data: tuple of ndarray
+        Data to process.
+    is_sel: tuple of bool
+        Of same length as data, indicate for each array if they are to be filtered or not.
+    npts: int
+        Number of points in the dataset
+    sampling: int
+        Number of points to sample. Must be striclty less than npts.
+    """
+    def __init__(self, data, is_sel, npts, sampling):
+        self.data = data
+        self.is_sel = is_sel
+        self.npts = npts
+        self.sampling = sampling
+        self.ks = np.random.permutation(npts)[:sampling]
+
+    @property
+    def nb_tests(self):
+        return self.sampling
+
+    def __iter__(self):
+        data = self.data
+        is_sel = self.is_sel
+        n = self.npts
+        sel = np.ones((n,), dtype=bool)
+        for k in self.ks:
+            sel[k] = False
+            yield (k, tuple(d[sel] if is_sel[i] else d for i, d in enumerate(data)))
+            sel[k] = True
+
+
+class LeaveKOutFolding(object):
+    """
+    Implement a variation on Leave-One-Out where the dataset is tiled into k folds. At each iteration, one fold is used
+    for testing, while the others are used for fitting.
+
+    Parameters
+    ----------
+    data: tuple of ndarray
+        Data to process.
+    is_sel: tuple of bool
+        Of same length as data, indicate for each array if they are to be filtered or not.
+    npts: int
+        Number of points in the dataset
+    folding: int
```

Comment:
folding should be renamed to a more informative `n_folds`.
Need to check naming with other parts and scikit-learn.